tensornetwork/network_components.py

```diff
@@ -844,12 +1040,1369 @@ def _load_edge(cls, edge_data: h5py.Group,
+  #TODO: this is recursive; use while loop for better performance
+  reachable_nodes = {node}
+  depth = 0
+  while True:
+    old_depth = depth
+    for edge in node.edges:
+
+      if edge.is_dangling():
+        continue
+      next_node = edge.node1 if edge.node1 is not node else edge.node2
+      if next_node in reachable_nodes:
+        continue
+      reachable_nodes.add(next_node)
+      depth += 1
+      node = next_node
+      continue
+
+    depth -= 1 if depth == old_depth else 0
+    if depth == 0:
+      break
+  return reachable_nodes
+
+
+def check_correct(nodes: Collection[BaseNode],
+                  check_connections: Optional[bool] = True) -> None:
+  """
+  Check if the network defined by `nodes` fulfills necessary
+  consistency relations.
+
+  Args:
+    nodes: A list of `BaseNode` objects.
+    check_connections: Check if the network is connected.
+  Returns:
+    None
+  Raises:
+    ValueError: If the network defined by `nodes` is not 
+      correctly structured.
+  """
+  for node in nodes:
+    for i, edge in enumerate(node.edges):
+      if edge.node1 is not node and edge.node2 is not node:
+        raise ValueError("Edge '{}' does not connect to node '{}'."
+                         "Edge's nodes: '{}', '{}'.".format(
+                             edge, node, edge.node1, edge.node2))
+
+      is_edge_node_consistent = False
+      if edge.node1 is node:
+        if edge.axis1 == i:
+          is_edge_node_consistent = True
+      if edge.node2 is node:
+        if edge.axis2 == i:
+          is_edge_node_consistent = True
+      if not is_edge_node_consistent:
+        raise ValueError(
+            "Edge '{}' does not point to '{}' on the correct axis. "
+            "Edge axes: {}, {}. Node axis: {}.".format(edge, node, edge.axis1,
+                                                       edge.axis2, i))
+  if check_connections:
+    check_connected(nodes)
+
+
+def check_connected(nodes: Collection[BaseNode]) -> None:
+  """
+  Check if all nodes in `nodes` are connected.
+  Args:
+    nodes: A list of nodes.
+  Returns:
+    None:
+  Raises:
+    ValueError: If not all nodes in `nodes` are connected.
+  """
+  if not set(nodes) <= reachable_deque(list(nodes)[0]):
+    raise ValueError("Non-connected graph")
+
+
+def get_all_nondangling(nodes: Collection[BaseNode]) -> Set[Edge]:
+  """Return the set of all non-dangling edges."""
+  edges = set()
+  for node in nodes:
+    edges |= node.get_all_nondangling()
+  return edges
+
+
+def get_all_nodes(edges: Collection[Edge]) -> Set[BaseNode]:
+  """Return the set of nodes connected to edges."""
+  nodes = set()
+  for edge in edges:
+    if edge.node1 is not None:
+      nodes |= {edge.node1}
+    if edge.node2 is not None:
+      nodes |= {edge.node2}
+
+  return nodes
+
+
+def get_all_edges(nodes: Collection[BaseNode]) -> Set[Edge]:
+  """Return the set of edges of all nodes."""
+  edges = set()
+  for node in nodes:
+    edges |= set(node.edges)
+  return edges
```

Comment:
Let's move all of these methods to a different file to prevent this file from being 2,400 lines long (!)

Reply:
this is a bit finicky due to the cyclic import issues. `__xor__, __or__ `and `__matmul__` require
`connect, contract_between, outer_product, _contract_trace` which in turn require
all the helpers `_flatten_edges_..., _remove_edges_...`.
I moved as many functions as possible for now into network_operations.py.

Reply:
I cut this down as much as possible for now. If it's still too long we need to find a good way of
breaking this apart without causing circular imports